enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,457 @@
+However, stretching L2 presents challenges at scale:
+- Network latency is not predictable
+- Traffic bottlenecks (e.g. single Top Of Rack)
+- L2 domain failures (e.g. broadcast storm)
+- Network management complexity
+
+Modern datacenters have adopted L3 protocols as part of their design and dynamic routing protocols help to
+build more resilient networks running at higher speed, than the traditional L2 domains, where the premise is
+that there is only a single path between two hosts.
+
+This enhancement helps to integrate OpenShift control plane into a Spine-Leaf architecture where each leaf would be its own Failure Domain and a smaller than usual L2 network, easier to maintain and more performant.
+
+### Traditional 3 layers architecture
+
+![image](on-prem-bgp-networking-stretched.png)
+
+In this design, the OpenShift nodes are placed on distinct zones (e.g. with OpenStack, Compute availability zones).
+However, the same OpenShift nodes are attached to the same physical or virtual network thus sharing the same network (L2) failure domain.
+
+### User Stories
+
+1. As a deployer of OpenShift, I would like to increase the API and Ingress availability of the cluster
+   so that OpenShift continues to function and continues to be reachable even in the event of a partial network outage.
+2. As a network infrastructure engineer, I would like to reduce network L2 domains and move the data center to
+   a modern L3 architecture so that I can reduce the complexity of the network infrastructure.
+3. As an OpenShift administrator, I want OpenShift to use my existing BGP infrastructure to publish API and Ingress endpoints.
+4. As an OpenShift administrator, I want my ingress traffic to not go through just one node, without having to use an external Load Balancer.
+
+
+### Goals
+
+The goal for this enhancement should be limited to providing a BGP-based solution for API and Ingress. We have [a separate enhancement proposal](https://github.com/openshift/enhancements/pull/1167)
+for allowing the control plane to sit in different failure domains.
+
+The minimal requirements includes:
+- Implement, test, document and deliver the feature for OpenStack
+- Highly available and dynamically routable (with BGP) load-balancing API access for external clients.
+- Highly available and dynamically routable (with BGP) access for default ingress.
+- Provide an interface for making use the BGP-based routing so other on-prem platforms like Baremetal, VMware vSphere, etc. can easily integrate the feature
+
+### Non-Goals
+
+The proposed BGP-based load-balancing requires an existing BGP infrastructure is already in place when deploying. For this reason the default load-balancing solution will remain Keepalived/HAproxy.
+Also, while this enhancement will be implemented in a way that allows all on-premise platforms to use it, right now we haven't planned the work to implement and test it for platforms other than OpenStack.
```

Comment:
this statement doesn't match the API.

Reply:
ack. Thanks for your review on the API PR, I'll take a look asap.